pkg/edit/builtins.go

```diff
@@ -563,60 +641,134 @@ func categorizeAlnum(r rune) int {
+//
+// Note that a word character must be to the right of pos.
+func (categorize categorizer) skipWordRight(buffer string, pos int) int {
+	if pos == len(buffer) {
+		return pos
+	}
+
+	r, _ := utf8.DecodeRuneInString(buffer[pos:])
 	cat := categorize(r)
+	// assert cat is not 0
+	return categorize.skipCatRight(cat, buffer, pos)
+}
+
+// Skip whitespace right from pos in buffer and return that position.
+func (categorize categorizer) skipWsRight(buffer string, pos int) int {
+	return categorize.skipCatRight(0, buffer, pos)
+}
+
+func (categorize categorizer) skipCatRight(cat int, buffer string, pos int) int {
+	right := strings.TrimLeftFunc(buffer[pos:], func(r rune) bool {
+		return categorize(r) == cat
+	})
+
+	return len(buffer) - len(right)
+}
+
+// Move the dot left one word, using the word flavor described by the
+// categorizer.
+func moveDotLeftGeneralWord(categorize categorizer, buffer string, dot int) int {
+	// skip trailing whitespaces left of dot
+	pos := categorize.skipWsLeft(buffer, dot)
 
 	// skip this word
-	skipCat(cat)
+	pos = categorize.skipWordLeft(buffer, pos)
 
-	return len(left)
+	return pos
 }
 
 // Move the dot right one word, using the word flavor described by the
 // categorizer.
 func moveDotRightGeneralWord(categorize categorizer, buffer string, dot int) int {
-	right := buffer[dot:]
-	skipCat := func(cat int) {
-		right = strings.TrimLeftFunc(right, func(r rune) bool {
-			return categorize(r) == cat
-		})
-	}
-
 	// skip leading whitespaces right of dot
-	skipCat(0)
+	pos := categorize.skipWsRight(buffer, dot)
 
 	// check whether any whitespace was skipped; if whitespace was
 	// skipped, then dot is already successfully moved to next
 	// non-whitespace run
-	if dot < len(buffer)-len(right) {
-		return len(buffer) - len(right)
+	if dot < pos {
+		return pos
 	}
 
 	// no whitespace was skipped, so we still have to skip to the next word
 
-	// get category of first rune
-	r, _ := utf8.DecodeRuneInString(right)
-	cat := categorize(r)
 	// skip this word
-	skipCat(cat)
+	pos = categorize.skipWordRight(buffer, pos)
 	// skip remaining whitespace
-	skipCat(0)
+	pos = categorize.skipWsRight(buffer, pos)
 
-	return len(buffer) - len(right)
+	return pos
+}
+
+// Transpose the words around the cursor, using the word flavor described
+// by the categorizer.
+func transposeGeneralWord(categorize categorizer, buffer string, dot int) (string, int) {
+	if len(strings.TrimFunc(buffer, func(r rune) bool { return categorize(r) == 0 })) == 0 {
+		// buffer contains only whitespace
+		return buffer, dot
+	}
+
+	// after skipping whitespace, find the end of the right word
+	pos := categorize.skipWsRight(buffer, dot)
+	var rightEnd int
+	if pos == len(buffer) {
+		// there is only whitespace to the right of the dot
+		rightEnd = categorize.skipWsLeft(buffer, pos)
+	} else {
+		rightEnd = categorize.skipWordRight(buffer, pos)
+	}
+	rightStart := categorize.skipWordLeft(buffer, rightEnd)
```

Comment:
IIUC this should be the same as `pos` if the `else` branch above was taken? If so, consider rewriting this as:

```go
var rightStart, rightEnd int
if pos == len(buffer) {
  ...
} else {
  rightEnd = ...
  rightLeft = pos
}
```

Reply:
No, if `dot` is in the middle of a word, then `pos` is `dot`, so we still need to skip word characters to the left to the find the left of the word. Will document this.